### PR TITLE
Mobile Backend: NHWC memory layout + XNNPACK integration. (#32509)

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Summary:
In order to improve CPU performance on floating-point models on mobile, this PR introduces a new CPU backend for mobile that implements the most common mobile operators with NHWC memory layout support through integration with XNNPACK.

XNNPACK itself, and this codepath, are currently only included in the build, but the actual integration is gated with USE_XNNPACK preprocessor guards.  This preprocessor symbol is intentionally not passed on to the compiler, so as to enable this rollout in multiple stages in follow up PRs.  This changeset will build XNNPACK as part of the build if the identically named USE_XNNPACK CMAKE variable, defaulted to ON, is enabled, but will not actually expose or enable this code path in any other way.

Furthermore, it is worth pointing out that in order to efficiently map models to these operators, some front-end method of exposing this backend to the user is needed.  The less efficient implementation would be to hook these operators into their corresponding native implementations, granted that a series of XNNPACK-specific conditions are met, much like how NNPACK is integrated with PyTorch today for instance.

Having said that, while the above implementation is still expected to outperform NNPACK based on the benchmarks I ran, the above integration would be leave a considerable gap between the performance achieved and the maximum performance potential XNNPACK enables, as it does not provide a way to compute and factor out one-time operations out of the inner most forward() loop.

The more optimal solution, and one we will  decide on soon, would involve either providing a JIT pass that maps nn operators onto these newly introduced operators, while allowing one-time calculations to be factored out, much like quantized mobile models.  Alternatively, new eager-mode modules can also be introduced that would directly call into these implementations either through c10 or some other mechanism, also allowing for decoupling of op creation from op execution.

This PR does not include any of the front end changes  mentioned above.  Neither does it include the mobile threadpool unification present in the original https://github.com/pytorch/pytorch/issues/30644.  Furthermore, this codepath seems to be faster than NNPACK in a good number of use cases, which can potentially allow us to remove NNPACK from aten to make the codebase a little simpler, granted that there is widespread support for such a move.

Regardless, these changes will be introduced gradually and in a more controlled way in subsequent PRs.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/32509

Test Plan:
Build: CI
Functionality: Not exposed

Reviewed By: dreiss

Differential Revision: D20069796

Pulled By: AshkanAliabadi

### CODE REVIEW

**Comment by AshkanAliabadi** [comment]
This patch is a re-land of #32509 which I reverted since it was breaking an iOS benchmark which seems to have been added in between I got green builds a week or two ago and the time I landed the diff.  This patch adds no new functionality except for fixing a bug which was causing the CI breakage.  I have marked that bug in the code.

File: ios/TestApp/benchmark/setup.rb [hunk 1]
```diff
@@ -63,7 +63,7 @@
     target.resources_build_phase.add_file_reference(config_file_ref, true)
 end
 puts "Linking static libraries..."
-libs = ['libc10.a', 'libclog.a', 'libnnpack.a', 'libeigen_blas.a', 'libcpuinfo.a', 'libpytorch_qnnpack.a', 'libtorch_cpu.a', 'libtorch.a']
+libs = ['libc10.a', 'libclog.a', 'libnnpack.a', 'libXNNPACK.a', 'libeigen_blas.a', 'libcpuinfo.a', 'libpytorch_qnnpack.a', 'libtorch_cpu.a', 'libtorch.a']
```
Comment (AshkanAliabadi) [inline ios/TestApp/benchmark/setup.rb]: This line was causing iOS benchmark to fail build.

**Comment by kurtamohler** [comment]
Recently I've found that I can no longer touch one source file on pytorch and then rebuild only that one part of the project. Instead, I have to clean and rebuild everything. I noticed that this issue doesn't happen if I set `USE_XNNPACK=0`. With that setting, I'm able to successfully rebuild only what I need.

Seeing that this commit is the most recent one that mentions XNNPACK, I'm guessing that the issue is coming from here.

**Comment by AshkanAliabadi** [comment]
> Recently I've found that I can no longer touch one source file in pytorch and then rebuild only that one part of the project. Instead, I have to clean and rebuild everything. I noticed that this issue doesn't happen if I set `USE_XNNPACK=0`. With that setting, I'm able to successfully rebuild only what I need.
> 
> Seeing that this commit is the most recent one that mentions XNNPACK, I'm guessing that the issue is coming from here.

@kurtamohler Would be happy to look into it if you can sen me repro steps.  What platform are you compiling for?  Seems I can modify a file and compile that one change on MacOS at least.

**Comment by kurtamohler** [comment]
To reproduce, first check out commit `aff1da5aaca50d89ba0d08a09ed5808571e99daf` from the master branch, making sure to sync the submodules with `$ git submodule sync --recursive && git submodule update --init --recursive
`. Clean and then do a debug build with `$ python setup.py clean` and then `$ DEBUG=1 python setup.py develop`. I have a few pytorch-related environment variables, not sure whether they make a difference or not:
```
export USE_CUDA=1
export USE_DISTRIBUTED=0
export USE_NCCL=0
export USE_MKLDNN=0
export USE_FBGEMM=0
export USE_NNPACK=0
export USE_QNNPACK=0
export CFLAGS="$CFLAGS -L$CONDA_PREFIX/lib"
```
This first build should succeed. Now run `$ touch aten/src/ATen/native/cpu/BinaryOpsKernel.cpp`, and then rerun `$ DEBUG=1 python setup.py develop`. This last command fails for me.

I'm using Ubuntu 18.04.3 LTS.

I'm also using conda with this environment:
```
# This file may be used to create an environment using:
# $ conda create --name <env> --file <this file>
# platform: linux-64
@EXPLICIT
https://conda.anaconda.org/pytorch/linux-64/magma-cuda101-2.5.1-1.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/mkl-include-2019.5-281.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/nvcc_linux-64-10.2-hc6a2c23_6.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/_libgcc_mutex-0.1-conda_forge.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ca-certificates-2019.11.28-hecc5488_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ld_impl_linux-64-2.33.1-h53a641e_7.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libgfortran-ng-7.3.0-hdf63c60_4.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libstdcxx-ng-9.2.0-hdf63c60_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/llvm-openmp-9.0.1-hc9558a2_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/_openmp_mutex-4.5-0_llvm.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libgomp-9.2.0-h24d8f2e_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libgcc-ng-9.2.0-h24d8f2e_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/mkl-2019.5-281.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/binutils_impl_linux-64-2.33.1-he1b5a44_7.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/bzip2-1.0.8-h516909a_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/expat-2.2.9-he1b5a44_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libffi-3.2.1-he1b5a44_1006.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libopenblas-0.3.7-h5ec1e0e_7.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libuv-1.34.0-h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ncurses-6.1-hf484d3e_1002.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ninja-1.10.0-hc9558a2_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/openssl-1.1.1d-h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/rhash-1.3.6-h14c3975_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/xz-5.2.4-h14c3975_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/yaml-0.2.2-h516909a_1.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/zlib-1.2.11-h516909a_1006.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/binutils_linux-64-2.33.1-h9595d00_16.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/gcc_impl_linux-64-7.3.0-hd420e75_4.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libblas-3.8.0-14_openblas.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libedit-3.1.20170329-hf8c457e_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libssh2-1.8.2-h22169c7_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/readline-8.0-hf8c457e_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/tk-8.6.10-hed695b0_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/binutils-meta-1.0.4-0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/gcc_linux-64-7.3.0-h553295d_16.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/gfortran_impl_linux-64-7.3.0-hdf63c60_4.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/gxx_impl_linux-64-7.3.0-hdf63c60_4.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/krb5-1.16.4-h2fd8d38_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libcblas-3.8.0-14_openblas.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/liblapack-3.8.0-14_openblas.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/sqlite-3.30.1-hcee41ef_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/c-compiler-1.0.4-h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/gfortran_linux-64-7.3.0-h553295d_16.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/gxx_linux-64-7.3.0-h553295d_16.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libcurl-7.65.3-hda55be3_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/python-3.6.7-h357f687_1006.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/attrs-19.3.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/certifi-2019.11.28-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/cmake-3.16.3-h28c56e5_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/cxx-compiler-1.0.4-hc9558a2_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/entrypoints-0.3-py36_1000.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/fortran-compiler-1.0.4-he991be0_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/mccabe-0.6.1-py_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/more-itertools-8.1.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/numpy-1.17.5-py36h95a1406_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/psutil-5.6.7-py36h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/py-1.8.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pycodestyle-2.5.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pycparser-2.19-py36_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pyflakes-2.1.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pyparsing-2.4.6-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pyyaml-5.3-py36h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/six-1.14.0-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/sortedcontainers-2.1.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/typing-3.6.4-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/wcwidth-0.1.8-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/cffi-1.13.2-py36h8022711_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/compilers-1.0.4-0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/hypothesis-5.3.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/packaging-20.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/scipy-1.4.1-py36h921218d_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/setuptools-45.1.0-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/flake8-3.7.9-py36_0.tar.bz2
https://repo.anaconda.com/pkgs/main/linux-64/wheel-0.33.6-py36_0.conda
https://conda.anaconda.org/conda-forge/linux-64/pip-20.0.2-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/importlib_metadata-1.4.0-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/inflect-4.0.0-py36_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pluggy-0.12.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/jaraco.itertools-5.0.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pytest-5.3.4-py36_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/zipp-2.1.0-py_0.tar.bz2
```

**Comment by AshkanAliabadi** [comment]
Thanks, will take a look tomorrow.

**Comment by kurtamohler** [comment]
Perfect, thank you! No rush or anything, since everything works fine if I set `USE_XNNPACK=0`.

**Comment by kurtamohler** [comment]
Welp, I was about to try to use git bisect to verify whether this commit caused my issue or not, but now I can't reproduce the issue using the steps I wrote above. ¯\_(ツ)_/¯ Nevermind, sorry to take up your time.

**Comment by AshkanAliabadi** [comment]
No worries, and thanks for letting me know.  What matters is that things
are OK.  :)

**Pull request closed without merging**